File: 01-Programming-Languages/Swift/Data-Structures/LinkedList/linked_list.swift
class Node {
    var value: Int
    var next: Node?
    
    init(value: Int) {
        self.value = value
        self.next = nil
    }
}

class LinkedList {
    var head: Node?
    
    func append(_ value: Int) {
        let newNode = Node(value: value)
        
        guard let head = head else {
            self.head = newNode
            return
        }
        
        var current = head
        while let next = current.next {
            current = next
        }
        current.next = newNode
    }
    
    func prepend(_ value: Int) {
        let newNode = Node(value: value)
        newNode.next = head
        head = newNode
    }
    
    func delete(_ value: Int) {
        guard let head = head else { return }
        
        if head.value == value {
            self.head = head.next
            return
        }
        
        var current = head
        while let next = current.next, next.value != value {
            current = next
        }
        
        if let next = current.next {
            current.next = next.next
        }
    }
    
    func display() {
        var current = head
        while let node = current {
            print("\(node.value) -> ", terminator: "")
            current = node.next
        }
        print("nil")
    }
}

let list = LinkedList()
list.append(1)
list.append(2)
list.append(3)
list.prepend(0)

print("List:")
list.display()

list.delete(2)
print("After deleting 2:")
list.display()

File: 01-Programming-Languages/Kotlin/Data-Structures/LinkedList/linked_list.kt
class Node(val value: Int) {
    var next: Node? = null
}

class LinkedList {
    var head: Node? = null

    fun append(value: Int) {
        val newNode = Node(value)
        if (head == null) {
            head = newNode
            return
        }

        var current = head!!
        while (current.next != null) {
            current = current.next!!
        }
        current.next = newNode
    }

    fun prepend(value: Int) {
        val newNode = Node(value)
        newNode.next = head
        head = newNode
    }

    fun delete(value: Int) {
        if (head?.value == value) {
            head = head?.next
            return
        }

        var current = head
        while (current?.next != null && current.next?.value != value) {
            current = current.next
        }
        current?.next = current?.next?.next
    }

    fun display() {
        var current = head
        while (current != null) {
            print("${current.value} -> ")
            current = current.next
        }
        println("null")
    }
}

fun main() {
    val list = LinkedList()
    list.append(1)
    list.append(2)
    list.append(3)
    list.prepend(0)

    println("List:")
    list.display()

    list.delete(2)
    println("After deleting 2:")
    list.display()
}